Add erosion/ dilation/ opening/ closing

K-Meech opened the pull request:
Closes https://github.com/haesleinhuepf/napari-segment-blobs-and-things-with-membranes/issues/32

This PR adds erosion, dilation, opening and closing form `skimage.morphology` - both the grayscale and binary versions. It also allows you to choose between using a square or disk footprint (both 2D and 3D versions)

Let me know if anything needs changing - happy to adapt this PR in any way!

K-Meech committed:
add grayscale and binary erosion, dilation, opening and closing

K-Meech committed:
add square or disk footprints

K-Meech committed:
fix imports

haesleinhuepf commented on napari_segment_blobs_and_things_with_membranes/__init__.py:
```diff
@@ -944,3 +987,139 @@ def sub_sample(image:"napari.types.ImageData", sample_x: int = 1, sample_y: int
     return np.squeeze(image)
+
+
+@register_function(menu="Segmentation post-processing > Grayscale erosion (scikit-image, nsbatwm)")
```
I presume Greyscale erosion is a filter (the name sounds like). Thus, it should be in a different menu. Furthermore, I'm wondering, is the underlying math identical with the `minimum_filter` ?

K-Meech replied:
Good point! I could move all the grayscale operations into `Filters/background removal`, and keep the binary ones under `Segmentation post-processing`? The only issue is that the grayscale operations can be quite useful for both `napari.types.ImageData` and  `napari.types.LabelData` e.g. grayscale erosion can be useful to erode a label image while retaining all the individual labels. Is there any way to allow a function to take both `napari.types.ImageData` and `napari.types.LabelsData` and potentially appear in two menus? It's not a big problem if not - it's easy to re-run connected components to get individual labels again.

Yes - I think grayscale erosion is equivalent to `minimum_filter`, as grayscale dilation is equivalent to `maximum_filter`. So these grayscale operations could be removed, as long as they were changed to use a disk footprint.

haesleinhuepf commented on napari_segment_blobs_and_things_with_membranes/__init__.py:
```diff
@@ -944,3 +987,139 @@ def sub_sample(image:"napari.types.ImageData", sample_x: int = 1, sample_y: int
+@jupyter_displayable_output(library_name='nsbatwm', help_url='https://www.napari-hub.org/plugins/napari-segment-blobs-and-things-with-membranes')
+@time_slicer
+def grayscale_dilation(labels: "napari.types.LabelsData", radius: int = 1,
+                       footprint_shape: FootprintShape = FootprintShape.square) -> "napari.types.LabelsData":
```
I'm not sure if the footprint_shape thingy is compatible with the Notebook generation.

haesleinhuepf reviewed (changes requested):
Hi @K-Meech 

awesome, thanks for working on this! I have two comments/questions below. On the second I will elaborate a bit in detail: After setting up a workflow using the new functions as [explained here](https://haesleinhuepf.github.io/BioImageAnalysisNotebooks/70_workflows/napari-assistant.html), users can generate Python code and notebooks from the workflow as [shown here](https://haesleinhuepf.github.io/BioImageAnalysisNotebooks/70_workflows/notebook_export.html).

In the case I tested, the generated code looked like this:
```
from skimage.io import imread
import napari_pyclesperanto_assistant as ncle  # version 0.22.1
import napari_segment_blobs_and_things_with_membranes as nsbatwm  # version 0.3.7
import napari
if 'viewer' not in globals():
    viewer = napari.Viewer()

image0_b = imread(
    "C:/Users/haase/mambaforge/envs/bio39/lib/site-packages/napari_pyclesperanto_assistant/data/blobs.tif")
viewer.add_image(image0_b, name='blobs')


# gauss otsu labeling

image1_gol = ncle.gauss_otsu_labeling(image0_b, 2.0)
viewer.add_labels(image1_gol, name='Result of gauss_otsu_labeling')

# grayscale erosion

image2_ge = nsbatwm.grayscale_erosion(image1_gol, 1, FootprintShape.square)
viewer.add_labels(image2_ge, name='Result of grayscale_erosion')
```

And caused this error in the notebook:
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 image2_ge = nsbatwm.grayscale_erosion(image1_gol, 1, FootprintShape.square)
      2 image2_ge

NameError: name 'FootprintShape' is not defined
```

I see two options how to deal with this:
* We need to find a way to add an import statement in the [code generator](https://github.com/haesleinhuepf/napari-workflows/blob/ffa3bdd298b99871a73cb06673881f635367f76c/src/napari_workflows/_workflow.py#L465), in case the FootPrintShape is used. I could imagine that this is quite complicated.
* The alternative is to not use Footprints at all and provide a radius parameter only and always use a Disc/Sphere shape. I think this is how nsbatwm worked so far. 

To make an educated decision: Do you see a use-case where different footprints are necessary to answer a scientific question?

Thanks again for your work! The code looks great :-)

K-Meech commented:
Thanks for taking a look @haesleinhuepf !

I'm happy to remove the Footprints and just use disk/sphere shape for all operations - I find that disk provides a better result than square most of the time anyway. I provided the option to switch between them, as many of the other filters in nsbatwm use square/cube shaped footprints. For example, the minimum filter, maximum filter, median filter, percentile filter... Each of these uses a `scipy.ndimage` function underneath that uses `footprint = np.ones((n, m))` e.g. here in the [minimum filter](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.minimum_filter.html). All these functions accept alternative footprints (and so could be changed to use disk), but this would change the result of all these operations - I wasn't sure if you would want to introduce such a breaking change?

haesleinhuepf commented:
> I find that disk provides a better result than square most of the time anyway

Me too actually! So let's go for them.



> other filters in nsbatwm use square/cube shaped footprints.

Good catch! I wasn't aware (even though I wrote them). Let's not deal with them now. I created a github issue #34 with a potential solution, including keeping backwards compatibility. No need to work on this from your side. But thanks for pointing out!

K-Meech committed:
Remove FootprintShape and add 3d image to test cases

K-Meech commented:
Thanks @haesleinhuepf - I've removed the footprint shape now, so everything I've added uses a disk/sphere footprint automatically. I also added a 3D image to the tests, so it tests with both the 2D disk and 3D sphere.

Do you have a preference for which menus these operations should go in? I added a comment above with some thoughts https://github.com/haesleinhuepf/napari-segment-blobs-and-things-with-membranes/pull/33#discussion_r1496260387

haesleinhuepf committed:
move greyscale operations to Filtering / background removal menu

haesleinhuepf committed:
renamed labels parameter to binary_image of binary_ functions

haesleinhuepf committed:
removed unnecessary import

haesleinhuepf commented:
Sorry, for the long silence @K-Meech . I just merged your new functions, moved some to the right menu, changed names and types ofthe greyscale functions and released a new version! 

Give it a try:

```
pip install napari-segment-blobs-and-things-with-membranes==0.3.8
```

Big thanks for your contribution!
